### PR TITLE
refactor: create SelfUserAPI in WireAPI package - WPB-8741

### PR DESCRIPTION
**Pull request by jullianm**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-8741" title="WPB-8741" target="_blank">TaskWPB-8741</a>  [iOS] Create SelfUserAPI
  </td></table>
  <br />
 

### Key points

SelfUserAPI has two endpoints: `getSelfUser` and `pushSupportedProtocols`. The implementation follows the same architecture pattern as the other APIs in the package. 

`getSelfUser` endpoint doesn't return a `supportedProtocols` object in v0, in v5 it does.
`pushSupportedProtocols` is only available from api version >= 5.

### Testing

Testing requests are properly created.
Testing both endpoints `getSelfUser`  and `pushSupportedProtocols` for failure and success.
Making sure calling the `pushSupportedProtocols` in api versions < v5 throws an error.

---

### Checklist

- [ ] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [ ] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [ ] Make sure you use the API for UI elements that support large fonts.
- [ ] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [ ] New UI elements have Accessibility strings for VoiceOver.

### CODE REVIEW

File: WireAPI/Sources/WireAPI/Models/SelfUser/SsoID.swift [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+/// The sso id of the self user
+
+public struct SsoID: Equatable {
```
Comment (caldrian) [inline WireAPI/Sources/WireAPI/Models/SelfUser/SsoID.swift]:
```suggestion
public struct SSOID: Equatable {
```

File: WireAPI/Tests/WireAPITests/Mocks/HTTPClientMock.swift [hunk 2]
```diff
@@ -97,7 +97,7 @@ final class HTTPClientMock: HTTPClient {
 
     convenience init(
         code: Int,
-        payload: Data
+        payload: Data?
```
Comment (caldrian) [inline WireAPI/Tests/WireAPITests/Mocks/HTTPClientMock.swift:100]: Why is this change necessary? We should avoid optionals wherever possible.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        let request = HTTPRequest(
+            path: "\(pathPrefix)/self",
+            method: .get
+        )
+        
+        let response = try await httpClient.executeRequest(request)
+
+        return try ResponseParser()
+            .success(code: 200, type: SelfUserV0.self)
+            .failure(code: 404, label: "not-found", error: SelfUserAPIError.selfUserNotFound)
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: I couldn't see this error response on swagger or in the existing code. Where did you find it?

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    }
+    
+    func pushSupportedProtocols(_ supportedProtocols: Set<SupportedProtocol>) async throws {
+        guard apiVersion >= .v5 else {
+            throw SelfUserAPIError.unsupportedEndpointForAPIVersion
+        }
+        
+        let encoder = JSONEncoder.defaultEncoder
+        let payload = SupportedProtocolsPayloadV5(supportedProtocols: supportedProtocols)
+        let body = try encoder.encode(payload)
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: This implementation is for v5 so I suggest to move this to SelfUserAPIV5. Instead in this method we could simply throw the error without checking the api version.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            path: "\(pathPrefix)/self/supported-protocols",
+            method: .put,
+            body: body
+        )
+        
+        let response = try await self.httpClient.executeRequest(request)
+
+        return try ResponseParser()
+            .success(code: 200)
+            .failure(code: 404, error: SelfUserAPIError.invalidRequest)
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: I also didn't find this error anywhere in swagger or the existing code.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+            .parse(response)
+        
+    }
+}
+
+struct SelfUserV0: Decodable, ToAPIModelConvertible {
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift:51]: According to Swagger the response model of `GET /self` is the same as some of the `/users` endpoints. This means that we can reuse the response model called `UserResponseV0` and avoid the duplication here.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+    let id: UUID
+    let locale: String
+    let managedBy: ManagedByV0?
+    let name: String
+    let phone: String?
+    let picture: [String]?
+    let qualifiedID: UserID
+    let service: ServiceResponseV0?
+    let ssoID: SsoIDV0?
+    let supportedProtocols: Set<SupportedProtocol>?
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: Supported protocols was only introduced in v4, which is also why `UserResponseV4` exists.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV4.swift [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+class SelfUserAPIV4: SelfUserAPIV0 {
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV4.swift]: Since the user response changed in v4 we should override the `GET /self` endpoint here to parse the different response.

File: WireAPI/Sources/WireAPI/Components/ResponseParser.swift [hunk 9]
```diff
@@ -63,20 +82,26 @@ struct ResponseParser<Success> {
...
         }
-
-        let failure = try decoder.decode(FailureResponse.self, from: data)
-        throw failure
+        
+        if let data {
+            let failure = try decoder.decode(FailureResponse.self, from: data)
+            throw failure
+        } else {
+            throw ParsingError.nothingToParse
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/Components/ResponseParser.swift]: I think rather than `nothingToParse` it would be more accurate to say the response had no parse result. Usually it should not happen but if it does then it would highlight to us that we have not handled the parsing correctly.

File: WireAPI/Sources/WireAPI/Models/SelfUser/SelfUser.swift [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+/// User profile for self
+
+public struct SelfUser: Equatable {
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/Models/SelfUser/SelfUser.swift:23]: I believe this duplicates the `User` model.

File: WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift [hunk 11]
```diff
@@ -0,0 +1,232 @@
...
+    // MARK: - Request generation
+
+    func testGetSelfUserRequest() async throws {
+        try await apiSnapshotHelper.verifyRequestForAllAPIVersions { sut in
+            _ = try await sut.getSelfUser()
+        }
+    }
+    
+    func testPushSupportedProtocolsRequest() async throws {
+        try await apiSnapshotHelper.verifyRequest(for: [.v5, .v6]) { sut in
```
Comment (johnxnguyen) [inline WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift]: Are we able to write `verifyRequest(for: .v5...)`? This way if there are new API versions added then they will be tested too.

File: WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift [hunk 12]
```diff
@@ -0,0 +1,232 @@
...
+            _ = try await sut.pushSupportedProtocols([.mls])
+        }
+    }
+    
+    // MARK: - Request unsupported endpoint
+    
+    func testPushSupportedProtocols_UnsupportedVersionError_V0() async throws {
```
Comment (johnxnguyen) [inline WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift]: What do you think about testing this also for v1, v2, and v3?

File: WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift [hunk 13]
```diff
@@ -0,0 +1,232 @@
...
+
+        // When
+        let sut = SelfUserAPIV5(httpClient: httpClient)
+
+        // Then
+        do {
+            try await sut.pushSupportedProtocols([.mls])
+        } catch let error {
+            throw error
+        }
```
Comment (johnxnguyen) [inline WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift]: This do catch is redundant.

File: WireAPI/Tests/WireAPITests/Mocks/HTTPClientMock.swift (hunk 2 above)
Comment (johnxnguyen) [reply WireAPI/Tests/WireAPITests/Mocks/HTTPClientMock.swift:100]: I believe because we encountered the first time a response has no payload data.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift (hunk 6 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift:51]: response model is slightly different from the UserResponseV0, for instance the SelfUser model has a SSOID object, no legalHoldStatus object and has an id property. Shouldn't we keep both models separated as, while looking similar, they still have these differences ?

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift (hunk 7 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: same comment as above, I can however create a SelfUserAPIV4 object that contains the supportedProtocols objects and remove the property from this one.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift (hunk 4 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: makes sense thanks 👍

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift (hunk 3 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: yes I haven't quite looked at the errors yet, I will do so and fix them

File: WireAPI/Tests/WireAPITests/Mocks/HTTPClientMock.swift (hunk 2 above)
Comment (jullianm) [reply WireAPI/Tests/WireAPITests/Mocks/HTTPClientMock.swift:100]: exactly, we need to handle use cases where API succeeds but no payload is sent back. this is the case for the pushSupportedProtocols API.

File: WireAPI/Sources/WireAPI/Components/ResponseParser.swift (hunk 9 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/Components/ResponseParser.swift]: makes sense thx

File: WireAPI/Sources/WireAPI/Models/SelfUser/SsoID.swift (hunk 1 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/Models/SelfUser/SsoID.swift]: fixed thx

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift (hunk 3 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: fixed o there are no explicit errors handling either in Swagger or legacy code with these endpoints, so I removed the explicit `.failure` handling from the ResponseParser and leave it to be decoded freely by the FailureResponse object.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift (hunk 4 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: done

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift (hunk 5 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: fixed, see comment above

File: WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift (hunk 11 above)
Comment (jullianm) [reply WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift]: done. I used a different approach though

File: WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift (hunk 12 above)
Comment (jullianm) [reply WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift]: done

File: WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift (hunk 13 above)
Comment (jullianm) [reply WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift]: fixed

File: WireAPI/Sources/WireAPI/Components/ResponseParser.swift [hunk 14]
```diff
@@ -35,26 +40,44 @@ struct ResponseParser<Success> {
         parseBlocks = []
     }
 
+    /// Success with output data
+
     func success<Payload: Decodable & ToAPIModelConvertible>(
         code: Int,
         type: Payload.Type
     ) -> ResponseParser<Success> where Payload.APIModel == Success {
+        precondition(200 ..< 300 ~= code, "Requires a valid success code: 2xx")
```
Comment (jullianm) [inline WireAPI/Sources/WireAPI/Components/ResponseParser.swift:49]: I think it would make sense to turn the code (Int) into an enum to ensure only success codes can be passed here. Same goes for failure. I will suggest these changes in my next PR.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV4.swift [hunk 15]
```diff
@@ -18,10 +18,76 @@
...
+            path: "\(pathPrefix)/self",
+            method: .get
+        )
+
+        let response = try await httpClient.executeRequest(request)
+
+        return try ResponseParser()
+            .success(code: 200, type: SelfUserV4.self)
+            .parse(response)
+    }
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV4.swift:38]: nitpick: it'd be nice if this was below the `apiVersion` property.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift (hunk 6 above)
Comment (johnxnguyen) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift:51]: ah yes you're right, my mistake. Yes they should be separate.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift [hunk 16]
```diff
@@ -0,0 +1,135 @@
...
+    }
+}
+
+struct SelfUserV0: Decodable, ToAPIModelConvertible {
+
+    let accentID: Int
+    let assets: [UserAsset]?
+    let deleted: Bool?
+    let email: String?
+    let expiresAt: UTCTimeMillis?
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift:57]: Swagger says this is `UTCTime` not `UTCTimeMillis`.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV4.swift [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+
+}
+
+struct SelfUserV4: Decodable, ToAPIModelConvertible {
+
+    let accentID: Int
+    let assets: [UserAsset]?
+    let deleted: Bool?
+    let email: String?
+    let expiresAt: UTCTimeMillis?
```
Comment (johnxnguyen) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV4.swift:48]: Swagger says this it `UTCTime` not `UTCTimeMillis`.

File: WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift [hunk 18]
```diff
@@ -0,0 +1,277 @@
...
+        // Then
+        await XCTAssertThrowsError {
+            // When
+            try await sut.getSelfUser()
+        }
+    }
+
+    // MARK: - V5
+
+    func testGetSelfUser_SuccessResponse_200_V5() async throws {
```
Comment (johnxnguyen) [inline WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift]: This endpoint didn't change in v5, but it did in v4. I think you only need to rename the test, resource name, and use the v4 api.

**Review by johnxnguyen — APPROVED** [review]

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift (hunk 16 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift:57]: looks like we need UTCTimeMillis as we end up with a decoding error using UTCTime as it expects to decode a date with fractional seconds. the format on Swagger is as follows: 2021-05-12T10:52:02.671Z

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift [hunk 19]
```diff
@@ -0,0 +1,135 @@
...
+        case managedBy = "managed_by"
+        case name, phone, picture
+        case qualifiedID = "qualified_id"
+        case service
+        case ssoID = "sso_id"
+        case teamID = "team"
+    }
+
+    func toAPIModel() -> SelfUser {
+        SelfUser(id: id,
```
Comment (samwyndham) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: Tiny thing but I guess we use a different style now:

```swift
SelfUser(
    id: id,
    qualifiedID: qualifiedID
    // ..
)
```

It would be nice if we could lint this 🤔

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV1.swift [hunk 20]
```diff
@@ -0,0 +1,27 @@
...
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+class SelfUserAPIV1: SelfUserAPIV0 {
```
Comment (samwyndham) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV1.swift:21]: Out of interest is this subclassing pattern for APIs something we are already using?

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV4.swift [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+        case name, phone, picture
+        case qualifiedID = "qualified_id"
+        case service
+        case ssoID = "sso_id"
+        case teamID = "team"
+        case supportedProtocols = "supported_protocols"
+    }
+
+    func toAPIModel() -> SelfUser {
+        SelfUser(id: id,
```
Comment (samwyndham) [inline WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV4.swift]: Same comment about formatting as earlier.

File: WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift [hunk 22]
```diff
@@ -0,0 +1,279 @@
...
+    func testPushSupportedProtocolsRequest() async throws {
+        let supportedVersions = APIVersion.v5.andNextVersions
+        try await apiSnapshotHelper.verifyRequest(for: supportedVersions) { sut in
+            _ = try await sut.pushSupportedProtocols([.mls])
+        }
+    }
+
+    // MARK: - Request unsupported endpoint
+
+    func testPushSupportedProtocols_UnsupportedVersionError_V0() async throws {
```
Comment (samwyndham) [inline WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift]: Personally I would do this for all versions where it is unsupported in a _single_ test. 

- As a form of documentation, it makes it easy to understand which API versions behave in the same way.
- It lowers the effort to test all API versions thoroughly.
- It saves a lot of boilerplate.

You can have an array of `APIVersion`s and build the `sut` using `SelfUserAPIBuilder`. While this kind of parameterized testing is not that common in iOS at the moment it will become common when we move to swift testing.

**Review by samwyndham — APPROVED** [review]
Nice work!

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift (hunk 19 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV0.swift]: done thx

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV4.swift (hunk 21 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV4.swift]: done thx

File: WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift (hunk 22 above)
Comment (jullianm) [reply WireAPI/Tests/WireAPITests/APIs/SelfUserAPI/SelfUserAPITests.swift]: Good point 👍  this way when we introduce a new API version, we make sure new versioned APIs are properly tested on the existing endpoints without having to write additional tests (as you mentioned), I did some changes just for SelfUser for now but it could also be nice to generalize and automatize this behavior to all APIs, we should keep this in mind.

File: WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV1.swift (hunk 20 above)
Comment (jullianm) [reply WireAPI/Sources/WireAPI/APIs/SelfUserAPI/SelfUserAPIV1.swift:21]: yes

**Review by samwyndham — APPROVED** [review]
